pkg/client/restclient/plugin.go

```diff
@@ -30,9 +30,22 @@ type AuthProvider interface {
 	// WrapTransport allows the plugin to create a modified RoundTripper that
 	// attaches authorization headers (or other info) to requests.
 	WrapTransport(http.RoundTripper) http.RoundTripper
+	// Login allows the plugin to initialize its configuration. It must not
+	// require direct user interaction.
+	Login() error
 }
 
-type Factory func() (AuthProvider, error)
+// Factory generates an AuthProvider plugin.
+//  clusterAddress is the address of the current cluster.
+//  config is the inital configuration for this plugin.
+//  persister allows the plugin to save updated configuration.
+type Factory func(clusterAddress string, config map[string]string, persister AuthProviderConfigPersister) (AuthProvider, error)
+
```

Comment:
> Would it make sense for the config to be a map[string]interface so that it could support lists of strings as well?

`map[string]interface{}` won't work well because we're trying to serialize the data out into json.

`map[string]string` works well, `map[string][]string` also works fine.  I don't have any objection to tweaking the type to `map[string][]string` in this pull or in a **very** close followup.  The datatype change would be a breaking serialization change for kubeconfig files, so we'd have to change it before anyone gets a chance to use it.